pkg/k8s/watchers/node.go

```diff
@@ -15,64 +15,99 @@
-				var valid, equal bool
-				if oldNode := k8s.ObjToV1Node(oldObj); oldNode != nil {
-					valid = true
-					if newNode := k8s.ObjToV1Node(newObj); newNode != nil {
-						oldNodeLabels := oldNode.GetLabels()
-						newNodeLabels := newNode.GetLabels()
-						if comparator.MapStringEquals(oldNodeLabels, newNodeLabels) {
-							equal = true
-						} else {
-							err := k.updateK8sNodeV1(oldNode, newNode)
-							k.K8sEventProcessed(metricNode, metricUpdate, err == nil)
+var (
+	// onceNodeInitStart is used to guarantee that only one function call of
+	// NodesInit is executed.
+	onceNodeInitStart sync.Once
+)
+
+// NodesInit initializes a k8s watcher for node events belonging to the node
+// where Cilium is running.
+func (k *K8sWatcher) NodesInit(k8sClient *k8s.K8sClient) {
+	onceNodeInitStart.Do(func() {
+		nodeStore, nodeController := informer.NewInformer(
+			cache.NewListWatchFromClient(k8sClient.CoreV1().RESTClient(),
+				"nodes", v1.NamespaceAll, fields.ParseSelectorOrDie("metadata.name="+nodeTypes.GetName())),
+			&v1.Node{},
+			0,
+			cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
+					var valid bool
+					if node := k8s.ObjToV1Node(obj); node != nil {
+						valid = true
+						if isCiliumTainted(node) {
+							k8sClient.ReMarkNodeReady()
+						}
+						err := k.updateK8sNodeV1(nil, node)
+						k.K8sEventProcessed(metricNode, metricCreate, err == nil)
+					}
+					k.K8sEventReceived(metricNode, metricCreate, valid, false)
+				},
+				UpdateFunc: func(oldObj, newObj interface{}) {
+					var valid, equal bool
+					if oldNode := k8s.ObjToV1Node(oldObj); oldNode != nil {
+						valid = true
+						if newNode := k8s.ObjToV1Node(newObj); newNode != nil {
+							if isCiliumTainted(newNode) {
+								k8sClient.ReMarkNodeReady()
+							}
+
+							oldNodeLabels := oldNode.GetLabels()
+							newNodeLabels := newNode.GetLabels()
+							if comparator.MapStringEquals(oldNodeLabels, newNodeLabels) {
+								equal = true
+							} else {
+								err := k.updateK8sNodeV1(oldNode, newNode)
+								k.K8sEventProcessed(metricNode, metricUpdate, err == nil)
+							}
 						}
 					}
-				}
-				k.K8sEventReceived(metricNode, metricUpdate, valid, equal)
+					k.K8sEventReceived(metricNode, metricUpdate, valid, equal)
+				},
 			},
-		},
-		nil,
-	)
+			nil,
+		)
 
-	k.blockWaitGroupToSyncResources(wait.NeverStop, nil, nodeController.HasSynced, k8sAPIGroupNodeV1Core)
-	go nodeController.Run(wait.NeverStop)
-	k.k8sAPIGroups.AddAPI(k8sAPIGroupNodeV1Core)
+		k.nodeStore = nodeStore
+
+		k.blockWaitGroupToSyncResources(wait.NeverStop, nil, nodeController.HasSynced, k8sAPIGroupNodeV1Core)
+		go nodeController.Run(wait.NeverStop)
+		k.k8sAPIGroups.AddAPI(k8sAPIGroupNodeV1Core)
+	})
 }
 
-func (k *K8sWatcher) updateK8sNodeV1(oldK8sNode, newK8sNode *slim_corev1.Node) error {
+// isCiliumTainted returns true if the given node has the Cilium Agent Not Ready
+// Node Taint.
+func isCiliumTainted(k8sNode *v1.Node) bool {
```

Comment:
a small unimportant nit about the name; I'd prefer something slightly more descriptive such as `hasAgentNotReadyTaint`.